### PR TITLE
perf: exp_rev_bits_len precompile

### PR DESCRIPTION
**Pull request by erabinov**

### CODE REVIEW

File: recursion/compiler/src/ir/utils.rs [hunk 1]
```diff
@@ -130,6 +130,41 @@ impl<C: Config> Builder<C> {
...
+
+        // Materialize the bit length as a Var.
+        let bit_len_var = bit_len.into().materialize(self);
+        // Get a pointer to the array of bits in the exponent.
+        let ptr = match power_bits {
+            Array::Dyn(ptr, _) => ptr,
+            _ => panic!("Expected a dynamic array"),
+        };
+
+        // Call the DslIR command ExpReverseBitsLen, which modifies the memory pointed to by `x_copy_arr_ptr`.
```
[the remark this answers is not in the document]
Comment (erabinov) [reply recursion/compiler/src/ir/utils.rs:157]:
```suggestion
        // Call the DslIR instruction ExpReverseBitsLen, which modifies the memory pointed to by `x_copy_arr_ptr`.
```

File: prover/src/lib.rs [hunk 2]
```diff
@@ -759,11 +759,11 @@ mod tests {
         let bytes = bincode::serialize(&wrapped_bn254_proof).unwrap();
 
         // Save the proof.
-        let mut file = File::create("proof-with-pis.bin").unwrap();
```
Comment (jtguibas) [inline prover/src/lib.rs:762]: make these .bin

File: recursion/compiler/src/ir/utils.rs [hunk 3]
```diff
@@ -130,6 +130,41 @@ impl<C: Config> Builder<C> {
...
+    /// A version of `exp_reverse_bits_len` that uses the ExpReverseBitsLen precompile.
+    pub fn exp_reverse_bits_len_fast(
+        &mut self,
+        x: Felt<C::F>,
+        power_bits: &Array<C, Var<C::N>>,
+        bit_len: impl Into<Usize<C::N>>,
+    ) -> Felt<C::F> {
+        // Instantiate an array of length one and store the value of x.
+        let mut x_copy_arr: Array<C, Felt<C::F>> = self.dyn_array(1);
+        self.set(&mut x_copy_arr, 0, x);
```
Comment (jtguibas) [inline recursion/compiler/src/ir/utils.rs:142]: nit: try to put space above with lines that have "//"

Comment (erabinov) [reply recursion/compiler/src/ir/utils.rs:142]: Rust analyzer gets rid of that white space because it's at the beginning of the function body.

Comment (jtguibas) [reply recursion/compiler/src/ir/utils.rs:142]: line below it

File: recursion/core/src/exp_reverse_bits/mod.rs [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+        // This only needs to be done for the first and last iterations.
+        // QUESTION: what happens when local.is_last and local.is_first are both on?
+        builder.recursion_eval_memory_access_single(
+            local.clk,
+            local.base_ptr,
+            &local.x,
+            local.x_mem_access_flag,
+        );
+
+        // The `base_ptr` column stays the same when not `is_last`.
```
Comment (jtguibas) [inline recursion/core/src/exp_reverse_bits/mod.rs:376]: Len also needs to stay the same

File: recursion/core/src/exp_reverse_bits/mod.rs [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+            local.ptr,
+            &local.current_bit,
+            memory_access,
+        );
+
+        // Constrain that the x_mem_access_flag is true when `is_first` or `is_last`.
+        builder.when(local.is_real).assert_eq(
+            local.x_mem_access_flag,
+            local.is_first + local.is_last - local.is_first * local.is_last,
+        );
```
Comment (jtguibas) [inline recursion/core/src/exp_reverse_bits/mod.rs:365]: nit: add space here

**Review by jtguibas — APPROVED** [review]

**Pull request merged**